Feature/vpn 3685 manage tunnel settings

trojanfoe opened the pull request:
## Ticket

JIRA-VPN-3685

## Description

We are migrating to a new configuration model where the configuration is owned by the daemon, which will allow it to start without the presence of a client.  New grpc methods have been added to set the configuration outside of the connect command and for backwards compatibility reasons we have left the original connect method in place and added a connect_v2 method.

If the configuration is changed while the tunnel is connected, then the tunnel is reconnected.

## Checklist:

- [ ] Changelog

## Screenshots (optional, if UI related)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)
<!-- Reviewable:end -->

trojanfoe committed:
VPN-3685: Add grpc event stream for settings.

Add the base service configuration manager.

trojanfoe committed:
Move other configuration types into `VpnServiceConfig`.

trojanfoe committed:
Delete TOML config when writing the JSON, not reading it.

trojanfoe committed:
Wire-up protobuf.

There are compilation errors, however.

trojanfoe committed:
Merge remote-tracking branch 'origin/develop' into feature/vpn-3685-manage-tunnel-settings

trojanfoe committed:
Send the service configuration one-at-a-time.

Also maintain backwards compatibility with older clients that do not support
the new configuration model.

trojanfoe committed:
Merge remote-tracking branch 'origin/develop' into feature/vpn-3685-manage-tunnel-settings

trojanfoe committed:
Add new command line options.

trojanfoe committed:
Merge remote-tracking branch 'origin/develop' into feature/vpn-3685-manage-tunnel-settings

trojanfoe committed:
Reconnect the tunnel when the config has changed.

trojanfoe commented:
In testing, I am seeing this message from the daemon:

```
2025-09-16T15:29:06.302276Z  INFO nym_gateway_client::client: Claiming more bandwidth with existing credentials. Stop the process now if you don't want that to happen.
2025-09-16T15:29:06.302332Z  WARN nym_gateway_client::client: Not enough bandwidth. Trying to get more bandwidth, this might take a while
2025-09-16T15:29:06.302374Z  INFO nym_gateway_client::client: The client is running in disabled credentials mode - attempting to claim bandwidth without a credential
```

We need to ensure that the credential mode is always on.

trojanfoe committed:
Save the configuration when the configuration changes.

pronebird reviewed (commented):
Yes, please pass `true` in place where we had a flag.

> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-:-O_LhOfn2PRifRdjsFOg:b75q0tm)* status: 0 of 28 files reviewed, all discussions resolved (waiting on @doums, @mrkiwi, @neacsu, and @rokas-ambrazevicius)



<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> @pronebird reviewed 5 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 28 files reviewed, 4 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpn-proto/proto/nym_vpn_service.proto` line 676 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_Li5qC4B8E9QdPcOTt:-O_Li5qC4B8E9QdPcOTu:b-qciz26) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-proto/proto/nym_vpn_service.proto#L676)):*
> ```proto
>   rpc SetEnableTwoHop(google.protobuf.BoolValue) returns (google.protobuf.Empty) {}
>   rpc SetNetstack(google.protobuf.BoolValue) returns (google.protobuf.Empty) {}
>   //rpc SetDisablePoissonRate(google.protobuf.BoolValue) returns (google.protobuf.Empty) {}
> ```

Please remove commented out code.
___
*[`nym-vpn-core/crates/nym-vpn-network-config/src/discovery.rs` line 357 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_Lhi1B7bn4CQOdQTYP:-O_Lhi1B7bn4CQOdQTYQ:b-j2wqef) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-core/crates/nym-vpn-network-config/src/discovery.rs#L357)):*
> ```XML
>             let fetched = Discovery::fetch(&discovery.network_name)
>                 .await
>                 .expect(&format!(
> ```

Expect does the same thing. please revert unrelated changes.
___
*[`nym-vpn-core/crates/nym-vpn-store/src/keys/wireguard/persistence/ephemeral.rs` line 12 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LjNOK0fepkEy1TkLb:-O_LjNOK0fepkEy1TkLc:btokapt) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-store/src/keys/wireguard/persistence/ephemeral.rs#L12)):*
> ```XML
> #[allow(dead_code)]
> #[derive(Default)]
> pub struct InMemEphemeralKeys {
> ```

This is not used anywhere, maybe we should consider deleting dead code? @neacsu  
___
*[`nym-vpn-core/crates/nym-vpn-proto/src/rpc_client.rs` line 76 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LibOl2pP39Qox5Y46:-O_LibOl2pP39Qox5Y47:bcji13d) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-proto/src/rpc_client.rs#L76)):*
> ```XML
>     pub async fn set_exit_point(&mut self, exit_point: ExitPoint) -> Result<()> {
>         let exit_node =
>             proto::ExitNode::try_from(exit_point.clone()).map_err(Error::InvalidRequest)?;
> ```

No need to clone `exit_point`


<!-- Sent from Reviewable.io -->

trojanfoe committed:
Fix change to `Dns` type in Tauri app.

trojanfoe committed:
Changes after review.

trojanfoe reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 4 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @pronebird, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpn-network-config/src/discovery.rs` line 357 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_Lhi1B7bn4CQOdQTYP:-O_LkwTrEXR2zHMxvCTa:b5k8klu) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-core/crates/nym-vpn-network-config/src/discovery.rs#L357)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Expect does the same thing. please revert unrelated changes.
</blockquote></details>

This was to squash a clippy warning:

```
warning: function call inside of `expect`
   --> crates\nym-vpn-network-config\src\discovery.rs:357:18
    |
357 |                   .expect(&format!(
    |  __________________^
358 | |                     "failed to fetch discovery for {}",
359 | |                     discovery.network_name
360 | |                 ));
    | |__________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#expect_fun_call
    = note: `#[warn(clippy::expect_fun_call)]` on by default
help: try
    |
357 ~                 .unwrap_or_else(|_| panic!("failed to fetch discovery for {}",
358 ~                     discovery.network_name));
    |

```
___
*[`nym-vpn-core/crates/nym-vpn-proto/proto/nym_vpn_service.proto` line 676 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_Li5qC4B8E9QdPcOTt:-O_LlPO11-KbMuUMO_x7:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-proto/proto/nym_vpn_service.proto#L676)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Please remove commented out code.
</blockquote></details>

Done.
___
*[`nym-vpn-core/crates/nym-vpn-proto/src/rpc_client.rs` line 76 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LibOl2pP39Qox5Y46:-O_LlcBE600X76AakEsh:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-proto/src/rpc_client.rs#L76)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

No need to clone `exit_point`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

trojanfoe reviewed (commented):
I was fairly certain I'd done that, so it's not clear what I missed.

> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-:-O_LmGJ46MMUtXovkvkT:bkcbjvu)* status: 3 of 29 files reviewed, 4 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @pronebird, and @rokas-ambrazevicius)



<!-- Sent from Reviewable.io -->

trojanfoe reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 3 of 29 files reviewed, 5 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @pronebird, and @rokas-ambrazevicius)

___
*[`nym-vpn-app/src-tauri/src/commands/tunnel.rs` line 54 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LmwN39zTRtCOSUN2c:-O_LmwN39zTRtCOSUN2d:b-iiskow) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-app/src-tauri/src/commands/tunnel.rs#L54)):*
> ```XML
>         .dns_server
>         .clone()
>         .map(|ip| nym_vpn_proto::proto::Dns { ip });
> ```

@rokas-ambrazevicius There has been a change to the `Dns` type in protobuf, making the IP address optional.  I was unable to see in the `npm-vpn-apple` codebase where the equivalent change would need to be made.  Could you take a look please?


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> @pronebird reviewed 2 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 8 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnc/src/cli.rs` line 72 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LnoWuCrs2AgADxd1P:-O_LnoWv5gudmkxG4ptY:b-s8skci) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnc/src/cli.rs#L72)):*
> ```XML
>     SetIpv6 {
>         /// Enable
>         #[arg(long, default_value = "false", action = ArgAction::SetTrue)]
> ```

Let's use on or off as per my example in private conversation
___
*[`nym-vpn-core/crates/nym-vpnc/src/cli.rs` line 80 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LnrTCAsPBY8e2xlSj:-O_LnrTCAsPBY8e2xlSk:buo9nmt) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnc/src/cli.rs#L80)):*
> ```XML
>         /// Enable
>         #[arg(long, default_value = "false", action = ArgAction::SetTrue)]
>         enabled: bool,
> ```

Same here
___
*[`nym-vpn-core/crates/nym-vpnc/src/cli.rs` line 87 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_Lns_iApjdX6r_SpMt:-O_Lns_iApjdX6r_SpMu:buo9nmt) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnc/src/cli.rs#L87)):*
> ```XML
>         /// Enable
>         #[arg(long, default_value = "false", action = ArgAction::SetTrue)]
>         enabled: bool,
> ```

Same here


<!-- Sent from Reviewable.io -->

neacsu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 8 unresolved discussions (waiting on @doums, @mrkiwi, @pronebird, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-store/src/keys/wireguard/persistence/ephemeral.rs` line 12 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LjNOK0fepkEy1TkLb:-O_Lo8682jC97gzV7bnI:blj6bnp) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-store/src/keys/wireguard/persistence/ephemeral.rs#L12)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

This is not used anywhere, maybe we should consider deleting dead code? @neacsu
</blockquote></details>

That code was the work I was doing but got diverted on having multiple wireguard keys, two for each gateway. The on disk correspondent is also not used, but has an `#[allow(unused)]`, so that's why it doesn't appear in warnings. Same macro should be used for this in memory structure (I must have forgot about putting one). The in memory one will probably be useful for the probe, which relies on in memory structures in other places of the codebase. Please don't remove it


<!-- Sent from Reviewable.io -->

trojanfoe committed:
Make nymc command line for set-ipv6, set-two-hop and set-netstack accept 'on|off`.

trojanfoe reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 4 of 29 files reviewed, 8 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @pronebird, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpnc/src/cli.rs` line 72 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LnoWuCrs2AgADxd1P:-O_LuKoL38Vw0rQQ30t9:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnc/src/cli.rs#L72)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Let's use on or off as per my example in private conversation
</blockquote></details>

Done.
___
*[`nym-vpn-core/crates/nym-vpnc/src/cli.rs` line 80 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LnrTCAsPBY8e2xlSj:-O_LuLXl5g2RRIuzW9cr:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnc/src/cli.rs#L80)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Same here
</blockquote></details>

Done.
___
*[`nym-vpn-core/crates/nym-vpnc/src/cli.rs` line 87 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_Lns_iApjdX6r_SpMt:-O_LuLuQEydc160EWm8k:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnc/src/cli.rs#L87)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Same here
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> @pronebird reviewed 1 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 4 of 29 files reviewed, 8 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpnc/src/main.rs` line 242 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LtpPP5PSGa9eRPFq5:-O_LtpPP5PSGa9eRPFq6:bwc9w57) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnc/src/main.rs#L242)):*
> ```XML
> 
> async fn set_entry_point(mut rpc_client: RpcClient, entry: CliEntry) -> Result<()> {
>     let Some(entry_point) = entry.entry_point()? else {
> ```

FYI this is nitpicking comment, this can be rewritten as:

```rs
if let Some(entry_point) = entry.entry_point()? {
  rpc_client.set_entry_point(entry_point).await?;
  Ok(())
} else [
  Err(anyhow!("You must specify ..."))
}
```

Subjectively this is a bit cleaner as we rely on implicit returns instead of magical `bail!`. Either way works. Thought to share what I was taught to do in the past many years working in rust :)


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 4 of 29 files reviewed, 8 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpnd/src/command_interface.rs` line 7 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LuYj21vji3CirFw8Y:-O_LuYj21vji3CirFw8Z:b-da2uf5) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/command_interface.rs#L7)):*
> ```XML
> 
> use futures::{StreamExt, stream::BoxStream};
> use nym_vpn_lib::gateway_directory::{EntryPoint, ExitPoint};
> ```

FYI nit: been trying to organize imports into groups separated with blank line:

1. Std
2. External crates
3. Nym crates, crate level imports

So maybe move that import somewhere closer to nym_vpn_lib_types


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> @pronebird reviewed 1 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 8 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/config.rs` line 16 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_Lvs-NBbg1-vxooGT7:-O_Lvs-NBbg1-vxooGT8:b6dspfs) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/config.rs#L16)):*
> ```XML
> use nym_vpnd_types::service::VpnServiceConfig;
> use serde::{Deserialize, Serialize, de::DeserializeOwned};
> use std::{
> ```

FYI let's organize imports as std, external, our stuff to avoid mixing them as they get sorted in a pretty weird way


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 8 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/config.rs` line 42 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_Lw3E86PuNUGgu8K8H:-O_Lw3E86PuNUGgu8K8I:b-tmc6g3) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/config.rs#L42)):*
> ```XML
> // Used for serializing the Percent type to 3 decimal places
> macro_rules! round_f64 {
>     ($f64:expr) => {{ ($f64 * 1000.0).round() / 1000.0 }};
> ```

FYI Ouch let me take a look at what can be done to avoid dealing with floats


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 9 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)



<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 10 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/config.rs` line 95 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LwhLq4WdGUJFQTWGN:-O_LwhLq4WdGUJFQTWGO:b-2jmr1y) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/config.rs#L95)):*
> ```XML
> #[allow(dead_code)]
> impl VpnServiceConfigManager {
>     #[allow(clippy::result_large_err)]
> ```

If `VpnServiceConfig` is too big, wrap it in `Box<..>`. We had stackoverflows on Windows in the past because of that.


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 11 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/config.rs` line 96 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_Lws7B7b2JwaaHCDeG:-O_Lws7B7b2JwaaHCDeH:bkhovyr) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/config.rs#L96)):*
> ```XML
> impl VpnServiceConfigManager {
>     #[allow(clippy::result_large_err)]
>     pub(super) fn new(network_config_dir: &Path) -> Result<Self> {
> ```

Just use `pub` in these methods. It's less verbose


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 12 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/config.rs` line 179 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LxI5R1wHGI2kbVaJO:-O_LxI5R1wHGI2kbVaJP:b-bhr786) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/config.rs#L179)):*
> ```XML
>     }
> 
>     #[allow(clippy::result_large_err)]
> ```

Either Box the entire config or Box individual fields inside of `VpnServiceConfig` that are too big.


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 14 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/config.rs` line 207 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LxSzABBDjURCuRh0G:-O_LxSzABBDjURCuRh0H:b-oha3jo) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/config.rs#L207)):*
> ```XML
>     }
> 
>     #[allow(clippy::result_large_err)]
> ```

Let's get rid of this..
___
*[`nym-vpn-core/crates/nym-vpnd/src/service/config.rs` line 222 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LxV7792GiKQSxvG5b:-O_LxV7792GiKQSxvG5c:b-5a4zck) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/config.rs#L222)):*
> ```XML
>     }
> 
>     #[allow(clippy::result_large_err)]
> ```

and this..


<!-- Sent from Reviewable.io -->

trojanfoe reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 14 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @pronebird, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpnd/src/command_interface.rs` line 7 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LuYj21vji3CirFw8Y:-O_LxZ0V3wAE6aHAhsgw:bww2azb) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/command_interface.rs#L7)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

FYI nit: been trying to organize imports into groups separated with blank line:

1. Std
2. External crates
3. Nym crates, crate level imports

So maybe move that import somewhere closer to nym_vpn_lib_types
</blockquote></details>

We need to use this `rustfmt.toml` then:

```
reorder_modules = true
reorder_imports = true
group_imports = "StdExternalCrate"
imports_granularity = "Crate"
```


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 5 of 29 files reviewed, 15 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/config.rs` line 116 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LxxUZ6NqawkwQ6tCF:-O_LxxUZ6NqawkwQ6tCG:bbw9wff) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/config.rs#L116)):*
> ```XML
>     pub(super) fn set_config(&mut self, config: VpnServiceConfig) {
>         self.config = config;
>         self.config_changed = true;
> ```

Should we not write to file in here?


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> @pronebird reviewed 1 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 6 of 29 files reviewed, 16 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/error.rs` line 37 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M0Wu35EFCg2M2Wm9b:-O_M0Wu35EFCg2M2Wm9c:b-zbgjo8) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/error.rs#L37)):*
> ```XML
> /// Not yet implemented.
> #[derive(Debug, thiserror::Error)]
> #[allow(dead_code)]
> ```

Let's not include dead code.


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> @pronebird reviewed 1 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 7 of 29 files reviewed, 17 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/vpn_service.rs` line 729 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M0orI1CRYgi24hNa4:-O_M0orI1CRYgi24hNa5:b-9savnn) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/vpn_service.rs#L729)):*
> ```XML
> 
>     async fn connect(&mut self) -> Result<()> {
>         if self.tunnel_state.is_connected() || self.tunnel_state.is_connecting() {
> ```

Please do not send disconnect to the tunnel, this will disengage firewall. Send connect, state machine will handle the rest


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> @pronebird reviewed 2 of 28 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 10 of 29 files reviewed, 18 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd-types-uniffi/src/service.rs` line 10 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M1v6Q1iMFFQbMLBnh:-O_M1v6Q1iMFFQbMLBni:b-sofvxg) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd-types-uniffi/src/service.rs#L10)):*
> ```XML
> 
> macro_rules! round_f32 {
>     ($float:expr) => {{ (($float * 1000.0).round() / 1000.0) as f32 }};
> ```

Let's not do this and instead store `u8` as we discussed in private.


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 10 of 29 files reviewed, 19 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-windows/winfw/src/winfw/version.h` line 6 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M2EDdBb_soFEr9WDM:-O_M2EDdBb_soFEr9WDN:bhay4k3) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-windows/winfw/src/winfw/version.h#L6)):*
> ```objc
> #define MINOR_VERSION 16
> #define PATCH_VERSION 0
> #define PRODUCT_VERSION "1.16.0"
> ```

Both version files are in gitignore because they are auto-generated during build time:

```
# Auto-generated version headers
wireguard/libwg/version.h
nym-vpn-windows/winfw/src/winfw/version.h
```

I don't think you have to commit changes to them


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 10 of 29 files reviewed, 18 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-network-config/src/discovery.rs` line 357 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_Lhi1B7bn4CQOdQTYP:-O_M2XEG9dlylaX9usuU:b-tx3b5n) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-core/crates/nym-vpn-network-config/src/discovery.rs#L357)):*
<details><summary><i>Previously, trojanfoe (Andy Duplain) wrote…</i></summary><blockquote>

This was to squash a clippy warning:

```
warning: function call inside of `expect`
   --> crates\nym-vpn-network-config\src\discovery.rs:357:18
    |
357 |                   .expect(&format!(
    |  __________________^
358 | |                     "failed to fetch discovery for {}",
359 | |                     discovery.network_name
360 | |                 ));
    | |__________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#expect_fun_call
    = note: `#[warn(clippy::expect_fun_call)]` on by default
help: try
    |
357 ~                 .unwrap_or_else(|_| panic!("failed to fetch discovery for {}",
358 ~                     discovery.network_name));
    |

```
</blockquote></details>

This is something new. Thanks!


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> @pronebird reviewed 1 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 10 of 29 files reviewed, 18 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)



<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 10 of 29 files reviewed, 18 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-app/src-tauri/src/commands/tunnel.rs` line 54 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LmwN39zTRtCOSUN2c:-O_M3APt4_wrjQ56dHZg:b-kw9xwg) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-app/src-tauri/src/commands/tunnel.rs#L54)):*
<details><summary><i>Previously, trojanfoe (Andy Duplain) wrote…</i></summary><blockquote>

@rokas-ambrazevicius There has been a change to the `Dns` type in protobuf, making the IP address optional.  I was unable to see in the `npm-vpn-apple` codebase where the equivalent change would need to be made.  Could you take a look please?
</blockquote></details>

This is for @doums 


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> @pronebird reviewed 3 of 28 files at r1, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 14 of 29 files reviewed, 19 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-api-client/src/types/mod.rs` line 17 at r5](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M3MQtEuJtLyW90GJv:-O_M3MQtEuJtLyW90GJw:b-uye2le) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/ef1c7037f037a51400af6971ed917ddf18640cc5/nym-vpn-core/crates/nym-vpn-api-client/src/types/mod.rs#L17)):*
> ```XML
> pub use platform::Platform;
> 
> pub use nym_contracts_common::{ContractsCommonError, NaiveFloat, Percent};
> ```

Do we need this error somewhere?


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 14 of 29 files reviewed, 20 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/mixnet/connect.rs` line 143 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M3RAD1IOdQwd_ZE4d:-O_M3RAD1IOdQwd_ZE4e:b-1x3q0m) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-core/crates/nym-vpn-lib/src/mixnet/connect.rs#L143)):*
> ```XML
>         .debug_config(debug_config)
>         .custom_shutdown(task_client)
>         .credentials_mode(setup_options.enable_credentials_mode)
> ```

keep it as `.credentials_mode(true)`


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> @pronebird reviewed 2 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 16 of 29 files reviewed, 20 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/mixnet/connect.rs` line 143 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M3RAD1IOdQwd_ZE4d:-O_M3cni21Dj88fA-vG5:b51l96x) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-core/crates/nym-vpn-lib/src/mixnet/connect.rs#L143)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

keep it as `.credentials_mode(true)`
</blockquote></details>

@neacsu do you object to fixing it or do you think we should keep the fields as is, and just pass `true` from above?


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> @pronebird reviewed 2 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 18 of 29 files reviewed, 21 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-lib-types/src/tunnel_state.rs` line 57 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M3tQ_ARKWeb6mAzlK:-O_M3tQ_ARKWeb6mAzlL:b-6ha2ft) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-lib-types/src/tunnel_state.rs#L57)):*
> ```XML
> impl TunnelState {
>     /// Returns true if the tunnel is in the connected state.
>     pub fn is_connected(&self) -> bool {
> ```

You can use matches! in place. However I don't think you need these helpers. We shouldn't need to check against individual states in most cases


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> @pronebird reviewed 2 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 20 of 29 files reviewed, 22 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-lib-types-uniffi/src/gateway_directory.rs` line 78 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M4RPM3UiW20H99v7P:-O_M4RPM3UiW20H99v7Q:b-no5n1h) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-lib-types-uniffi/src/gateway_directory.rs#L78)):*
> ```XML
>     Gateway { identity: NodeIdentity },
>     Location { location: String },
>     Random,
> ```

Ouch we gotta put that into release!


<!-- Sent from Reviewable.io -->

neacsu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 20 of 29 files reviewed, 22 unresolved discussions (waiting on @doums, @mrkiwi, @pronebird, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/mixnet/connect.rs` line 143 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M3RAD1IOdQwd_ZE4d:-O_M4JDn2hCqbR91RaCc:blt399h) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-core/crates/nym-vpn-lib/src/mixnet/connect.rs#L143)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

@neacsu do you object to fixing it or do you think we should keep the fields as is, and just pass `true` from above?
</blockquote></details>

I think we should specify `.credentials_mode(true)` since the nym repo is not yet ready ditch this configuration and currently has this flag as `false` by default on the sdk side, from what I recall


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 20 of 29 files reviewed, 22 unresolved discussions (waiting on @doums, @mrkiwi, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-lib-types-uniffi/src/gateway_directory.rs` line 78 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M4RPM3UiW20H99v7P:-O_M5H6rC3vLGMZMTXl4:b-34g7o2) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-lib-types-uniffi/src/gateway_directory.rs#L78)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Ouch we gotta put that into release!
</blockquote></details>

Extracted to https://github.com/nymtech/nym-vpn-client/pull/3446


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 20 of 29 files reviewed, 22 unresolved discussions (waiting on @doums, @mrkiwi, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-network-config/src/system_messages.rs` line 227 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M5YRu3NqUuCHgaqjc:-O_M5YRu3NqUuCHgaqjd:b-378hwk) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-core/crates/nym-vpn-network-config/src/system_messages.rs#L227)):*
> ```XML
>         }"#;
>         let parsed: SystemMessageResponse = serde_json::from_str(json).unwrap();
>         let message = SystemMessage::try_from(parsed).unwrap();
> ```

FYI SMH, how did it even compile before... :scream: 


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 20 of 29 files reviewed, 23 unresolved discussions (waiting on @doums, @mrkiwi, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-proto/proto/nym_vpn_service.proto` line 73 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M5hTX0E5dH7hNr8dS:-O_M5hTX0E5dH7hNr8dT:b59iny9) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-proto/proto/nym_vpn_service.proto#L73)):*
> ```proto
> }
> 
> message Percent { optional float percent = 1; }
> ```

Let's get rid of this if we can...


<!-- Sent from Reviewable.io -->

pronebird reviewed (changes requested):
> @pronebird reviewed 2 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 22 of 29 files reviewed, 23 unresolved discussions (waiting on @doums, @mrkiwi, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd-types-uniffi/src/service.rs` line 41 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M5rFt3HNqW_NeogTQ:-O_M5rFt3HNqW_NeogTR:bvx4woa) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd-types-uniffi/src/service.rs#L41)):*
> ```XML
>             min_mixnode_performance: config
>                 .min_mixnode_performance
>                 .map(|p| round_f32!(p.naive_to_f64() as f32)),
> ```

Let's switch to u8 here too


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> @pronebird reviewed 2 of 28 files at r1.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 24 of 29 files reviewed, 23 unresolved discussions (waiting on @doums, @mrkiwi, @rokas-ambrazevicius, and @trojanfoe)



<!-- Sent from Reviewable.io -->

rokas-ambrazevicius commented:
> > _[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)_ status: 3 of 29 files reviewed, 5 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @pronebird, and @rokas-ambrazevicius)
> 
> _[`nym-vpn-app/src-tauri/src/commands/tunnel.rs` line 54 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LmwN39zTRtCOSUN2c:-O_LmwN39zTRtCOSUN2d:b-iiskow) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-app/src-tauri/src/commands/tunnel.rs#L54)):_
> 
> > ```
> >         .dns_server
> >         .clone()
> >         .map(|ip| nym_vpn_proto::proto::Dns { ip });
> > ```
> 
> @rokas-ambrazevicius There has been a change to the `Dns` type in protobuf, making the IP address optional. I was unable to see in the `npm-vpn-apple` codebase where the equivalent change would need to be made. Could you take a look please?

Don't worry about it, we will make the apps conform during the integration 👍

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 24 of 29 files reviewed, 23 unresolved discussions (waiting on @doums, @mrkiwi, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd/src/command_interface.rs` line 7 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_LuYj21vji3CirFw8Y:-O_M7y-dA4DAMcnZkRgr:biovc9u) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/command_interface.rs#L7)):*
<details><summary><i>Previously, trojanfoe (Andy Duplain) wrote…</i></summary><blockquote>

We need to use this `rustfmt.toml` then:

```
reorder_modules = true
reorder_imports = true
group_imports = "StdExternalCrate"
imports_granularity = "Crate"
```
</blockquote></details>

I think that's a good suggestion. Let's explore that separately.


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
I left a comment somewhere on this review board

> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-:-O_M82W32W_7RULPIRzQ:b4i1yjt)* status: 24 of 29 files reviewed, 23 unresolved discussions (waiting on @doums, @mrkiwi, @rokas-ambrazevicius, and @trojanfoe)



<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> @pronebird reviewed 3 of 3 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: all files reviewed (commit messages unreviewed), 18 unresolved discussions (waiting on @doums, @mrkiwi, @rokas-ambrazevicius, and @trojanfoe)



<!-- Sent from Reviewable.io -->

trojanfoe committed:
Drop `Percent` in `VpnServiceConfig`.

pronebird reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 23 of 29 files reviewed, 18 unresolved discussions (waiting on @doums, @mrkiwi, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpn-lib-types-uniffi/src/gateway_directory.rs` line 78 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M4RPM3UiW20H99v7P:-O_MDv8dC7v6HVtpyv0l:b-n6khor) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-lib-types-uniffi/src/gateway_directory.rs#L78)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Extracted to https://github.com/nymtech/nym-vpn-client/pull/3446
</blockquote></details>

Please rebase


<!-- Sent from Reviewable.io -->

trojanfoe committed:
Avoid disconnecting before reconnecting.

trojanfoe reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 23 of 29 files reviewed, 17 unresolved discussions (waiting on @doums, @mrkiwi, @pronebird, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/vpn_service.rs` line 729 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M0orI1CRYgi24hNa4:-O_MIYKrDAVszKiqLYTi:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/vpn_service.rs#L729)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Please do not send disconnect to the tunnel, this will disengage firewall. Send connect, state machine will handle the rest
</blockquote></details>

Done.
___
*[`nym-vpn-core/crates/nym-vpnd-types-uniffi/src/service.rs` line 10 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M1v6Q1iMFFQbMLBnh:-O_MIdjf6DWKUTl6K1As:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd-types-uniffi/src/service.rs#L10)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Let's not do this and instead store `u8` as we discussed in private.
</blockquote></details>

Done.
___
*[`nym-vpn-windows/winfw/src/winfw/version.h` line 6 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M2EDdBb_soFEr9WDM:-O_MIgAN1PxN46gO4vEG:bf3op49) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-windows/winfw/src/winfw/version.h#L6)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Both version files are in gitignore because they are auto-generated during build time:

```
# Auto-generated version headers
wireguard/libwg/version.h
nym-vpn-windows/winfw/src/winfw/version.h
```

I don't think you have to commit changes to them
</blockquote></details>

I guess they need adding to `.gitignore` then?


<!-- Sent from Reviewable.io -->

trojanfoe committed:
Remove config toml file when creating the config manager.

trojanfoe reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 23 of 29 files reviewed, 17 unresolved discussions (waiting on @doums, @mrkiwi, @pronebird, and @rokas-ambrazevicius)

___
*[`nym-vpn-windows/winfw/src/winfw/version.h` line 6 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M2EDdBb_soFEr9WDM:-O_MVeIk4wJU23kxRFBc:b3m3yka) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-windows/winfw/src/winfw/version.h#L6)):*
<details><summary><i>Previously, trojanfoe (Andy Duplain) wrote…</i></summary><blockquote>

I guess they need adding to `.gitignore` then?
</blockquote></details>

Ah it is in the `.gitignore` file; so why isn't it being ignored?


<!-- Sent from Reviewable.io -->

trojanfoe committed:
Drop `SetConfigError` and `ConfigSetupError` instead.

trojanfoe reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 22 of 29 files reviewed, 17 unresolved discussions (waiting on @doums, @mrkiwi, @pronebird, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/error.rs` line 37 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M0Wu35EFCg2M2Wm9b:-O_MYGX-CC8mpgrwGfrB:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpnd/src/service/error.rs#L37)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Let's not include dead code.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

trojanfoe committed:
Merge remote-tracking branch 'origin/develop' into feature/vpn-3685-manage-tunnel-settings

trojanfoe reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 20 of 29 files reviewed, 17 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @pronebird, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpn-api-client/src/types/mod.rs` line 17 at r5](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M3MQtEuJtLyW90GJv:-O_Md8Zs6sG6EaHOulYo:bp12dxe) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/ef1c7037f037a51400af6971ed917ddf18640cc5/nym-vpn-core/crates/nym-vpn-api-client/src/types/mod.rs#L17)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Do we need this error somewhere?
</blockquote></details>

Not any longer; removed.


<!-- Sent from Reviewable.io -->

trojanfoe committed:
Always set credential mode and remove unused error imports.

trojanfoe reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 20 of 29 files reviewed, 17 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @pronebird, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/mixnet/connect.rs` line 143 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M3RAD1IOdQwd_ZE4d:-O_Me5Lw2yu178AJC6xB:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d0bc7914ee6ecc4082a6da206770bd4b96d739c6/nym-vpn-core/crates/nym-vpn-lib/src/mixnet/connect.rs#L143)):*
<details><summary><i>Previously, neacsu (Bogdan-Ștefan Neacşu) wrote…</i></summary><blockquote>

I think we should specify `.credentials_mode(true)` since the nym repo is not yet ready ditch this configuration and currently has this flag as `false` by default on the sdk side, from what I recall
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

trojanfoe committed:
Remove `TunnelState` helpers.

trojanfoe reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 16 of 29 files reviewed, 17 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @pronebird, and @rokas-ambrazevicius)

___
*[`nym-vpn-core/crates/nym-vpn-lib-types/src/tunnel_state.rs` line 57 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M3tQ_ARKWeb6mAzlK:-O_MfKU23YfYnj4q5Vxt:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-lib-types/src/tunnel_state.rs#L57)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

You can use matches! in place. However I don't think you need these helpers. We shouldn't need to check against individual states in most cases
</blockquote></details>

Done.
___
*[`nym-vpn-core/crates/nym-vpn-lib-types-uniffi/src/gateway_directory.rs` line 78 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M4RPM3UiW20H99v7P:-O_MeNel3R6poVGo_WMz:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-core/crates/nym-vpn-lib-types-uniffi/src/gateway_directory.rs#L78)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Please rebase
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

pronebird committed:
Introduce TargetState and change connect_v2 and reconnect semantics

- Connect will only connect if the tunnel had not been connected
- Reconnect will only reconnect if the tunnel had been connected

pronebird reviewed (changes requested):
> @pronebird reviewed 3 of 6 files at r6, 1 of 2 files at r10, 4 of 5 files at r11, 1 of 3 files at r12.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 18 of 29 files reviewed, 9 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-core/crates/nym-vpnd/src/service/config.rs` line 298 at r12](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_MnU5KBetqy5BOeERN:-O_MnU5LEDEdv-BoH2Jr:b-hs6nhy) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/2ed2beb734c423a6b89cbe79697009c1a04e86c8/nym-vpn-core/crates/nym-vpnd/src/service/config.rs#L298)):*
> ```XML
> impl Drop for VpnServiceConfigManager {
>     fn drop(&mut self) {
>         self.write_to_file();
> ```

Let's not write to file if the config has not been changed


<!-- Sent from Reviewable.io -->

trojanfoe committed:
Remove overuse of `min(100)` to validate per percent values.

pronebird reviewed (changes requested):
> @pronebird reviewed 1 of 6 files at r6, 1 of 2 files at r10, 1 of 3 files at r12.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 17 of 29 files reviewed, 7 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)

___
*[`nym-vpn-windows/winfw/src/winfw/version.h` line 6 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-O_M2EDdBb_soFEr9WDM:-O_Mnyk0Ewo3vobkZQej:b-5v2qru) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/0104fda8380a6e3c752f8090610b20b436281797/nym-vpn-windows/winfw/src/winfw/version.h#L6)):*
<details><summary><i>Previously, trojanfoe (Andy Duplain) wrote…</i></summary><blockquote>

Ah it is in the `.gitignore` file; so why isn't it being ignored?
</blockquote></details>

Ok leave it.. But yeah it's weird they aren't ignored..


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> @pronebird reviewed 1 of 3 files at r9, 1 of 5 files at r11, 1 of 3 files at r12.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443)* status: 20 of 29 files reviewed, 4 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)



<!-- Sent from Reviewable.io -->

pronebird committed:
Rely on target state

trojanfoe committed:
FIx errors... that appeared from nowhere?

pronebird committed:
Fix elephant in the room

trojanfoe committed:
Fix clippy error, that's blocking the build.

pronebird committed:
Fix pinmut

pronebird committed:
Remove unused box

pronebird committed:
Feed initial settings into state machine

pronebird committed:
fix: clippy

trojanfoe committed:
Read tunnel settings when creating the vpn service.

trojanfoe committed:
Merge branch 'feature/vpn-3685-manage-tunnel-settings' of https://github.com/nymtech/nym-vpn-client into feature/vpn-3685-manage-tunnel-settings

pronebird committed:
state_machine: handle connect in connecting state to force reconnect

trojanfoe committed:
Add service config v2, with migration from v1 and TOML.

trojanfoe committed:
Merge branch 'feature/vpn-3685-manage-tunnel-settings' of https://github.com/nymtech/nym-vpn-client into feature/vpn-3685-manage-tunnel-settings

pronebird committed:
Maintain legacy behavior of connect method

trojanfoe committed:
Minor comment changes.

trojanfoe committed:
We don't need `Default` for the external config types.

pronebird reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @pronebird reviewed 5 of 7 files at r13, 5 of 5 files at r15, 1 of 4 files at r17.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3443#-:-O_R8yTw9nG0iQbFDPhc:bnfp4nl)* status: 27 of 30 files reviewed, 2 unresolved discussions (waiting on @doums, @mrkiwi, @neacsu, @rokas-ambrazevicius, and @trojanfoe)



<!-- Sent from Reviewable.io -->
